src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -937,23 +948,32 @@ int MPIDI_OFI_mpi_free_mem(void *ptr)
 
 /* ---- static functions for vni contexts ---- */
 static int create_vni_domain(struct fid_domain **p_domain, struct fid_av **p_av,
-                             struct fid_cntr **p_cntr);
+                             struct fid_cntr **p_cntr, int nic);
 static int create_cq(struct fid_domain *domain, struct fid_cq **p_cq);
 static int create_sep_tx(struct fid_ep *ep, int idx, struct fid_ep **p_tx,
-                         struct fid_cq *cq, struct fid_cntr *cntr);
-static int create_sep_rx(struct fid_ep *ep, int idx, struct fid_ep **p_rx, struct fid_cq *cq);
-static int try_open_shared_av(struct fid_domain *domain, struct fid_av **p_av);
-static int create_rma_stx_ctx(struct fid_domain *domain, struct fid_stx **p_rma_stx_ctx);
+                         struct fid_cq *cq, struct fid_cntr *cntr, int nic);
+static int create_sep_rx(struct fid_ep *ep, int idx, struct fid_ep **p_rx, struct fid_cq *cq,
+                         int nic);
+static int try_open_shared_av(struct fid_domain *domain, struct fid_av **p_av, int nic);
 static int open_non_shared_av(struct fid_domain *p_domain, struct fid_av **p_av);
 
-static int create_vni_context(int vni)
+/* This function creates a vni context which includes all of the OFI-level objects needed to
+ * initialize OFI (e.g. domain, address vector, endpoint, etc.). This function takes two arguments:
+ *
+ * vni - The VNI index within a nic to use when assigning the OFI information being created.
+ * nic - The NIC that should be used when setting up the OFI interfaces.
+ *
+ * Each nic will restart its vni indexing. This allows each VNI to use any nic if desired.
+ */
+static int create_vni_context(int vni, int nic)
 {
     int mpi_errno = MPI_SUCCESS;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_CREATE_VNI_CONTEXT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_CREATE_VNI_CONTEXT);
 
-    struct fi_info *prov_use = MPIDI_OFI_global.prov_use;
+    struct fi_info *prov_use = MPIDI_OFI_global.prov_use[nic];
```

Comment:
I see. It is the same provider, but of will return one for each nic?

Reply:
In MPICH, we call each of these a provider, but they have different NIC addresses under the hood so we have to have multiple copies even though they will all provide the same features.